GANDLF/utils/gandlf_logger.py

```diff
@@ -0,0 +1,44 @@
+import logging
+import yaml
+from pathlib import Path
+from importlib import resources
+
+
+def gandlf_logger_setup(config_path="logging_config.yaml") -> logging.Logger:
+    """
+    It sets up the logger. Read from logging_config.
+    Args:
+        logger_name (str): logger name, the name should be the same in the logging_config
+        config_path (str): file path for the configuration
+    Returns:
+        logging.Logger
+    """
+
+    # create dir for storing the messages
+    current_dir = Path.cwd()
+    directory = Path.joinpath(current_dir, "tmp/gandlf")
```

Comment:
This can be prone to breaks. For example, if an HPC admin has done an installation of GaNDLF into an environment that requires elevated privileges to make changes (a very common form of structuring "frozen" virtual environments in HPCs), then logging would break. A much more robust solution would be to generate the logs in the `output_dir` of the `parameters` dict, which gets initialized during both training [[ref](https://github.com/mlcommons/GaNDLF/blob/master/GANDLF/compute/training_loop.py#L235)] and inference [[ref](https://github.com/mlcommons/GaNDLF/blob/master/GANDLF/inference_manager.py#L37)]. This (or a similar) option will almost invariably be present for any GaNDLF process, so we can use that. In case there isn't any such option, I would vote to flush to console rather than putting it into this directory. 

Thoughts @VukW?

Reply:
What we do have in params?
`anonymizer`: output folder
`collect_stats`: output folder
`config_generator`: output folder
`construct_csv`: output file
`debug_info`: none
`deploy`: output folder
`generate_metrics`: output file
`optimize_model`: None? (there is a input file `model.pth.tar`, where is it located? maybe we can save logs nearby?)
`patch_miner`: output folder
`preprocess`: output folder
`recover_config`: output file
`run`: output folder (`--model-dir` arg)
`split-csv`: output folder
`verify-install`: None

So we don't have a fixed path for all the commands altogether - it needs to be calculated dynamically. The issue is we fix a logging file location in the `logging_config`..

Maybe we can do a workaround? Instead of passing a different output folder params for different commands, make a common `--output-dir` param for the whole `gandlf` cli command, and pass it to the scripts if it requires such a dir? like,
```
gandlf --output-folder ./output run ...
gandlf --output-folder . recover-config ...
```
If so, then we can make say
- make a current path as output by default (`.`) so `recover-config`, `debug-info` and others to store logs right here
- for all other commands (like run, deploy etc) require a `--output-folder` param to be filled explicitly


Another possible solution is to store logs in temporary dir instead `/tmp/gandlf` for unix, `%TEMP%` for windows
https://docs.python.org/3/library/tempfile.html
https://stackoverflow.com/a/43418319

In both the solutions we have to define a log location dynamically while `gandlf_logger_setup` instead of hardcoding the path in `logging_config.yaml`

Personally I'm in favor of the second solution. How about you?

Third possible solution (a bit ugly as we setup logging __before__ normal commands args are parsed): unify all output dir args to the same name, say, `output-dir` so cli looks like this:
```
gandlf run --output-dir ./output/ ....
gandlf deploy --output-dir ./output/ ...
gandlf debug-info  # no output-dir here
```
Thus, while user enters `gandlf` cli, we extract this param from the context (if present) and use it as log folder destination. Instead, just use `.`

Reply:
Hi @VukW, how are you? 
I agree with you, the second solution is the best.
I will try to implement it. 
Thanks for your suggestions.!!!

Reply:
Thanks for pointing out the discrepancy for `optimize_model`, @VukW! I have added this something we need to fix in #895.

Reply:
> Thanks for pointing out the discrepancy for `optimize_model`, @VukW! I have added this something we need to fix in #895.

Update: this should get fixed with #896.

Reply:
Hey @sarthakpati , that's great. However we still have `generate_metrics`, `construct_csv` and `recover_config` that do not have an output folder (just a path to output file). How do you think, can we unify this also to have an output folder? + we have `debug-info` + `verify-install` that do not have the output folder in their concept - so there's a question where to save gandlf logs for them

Reply:
I see 2 solutions:

1. How about this: let's default the location of the logs to `None`, in which case it flushes the logs to the console. In case the location is a directory, it writes the log there. This way, we don't have to change the invocation of `generate_metrics`, `construct_csv` and `recover_config` (which are relatively quick-running processes anyway).
2. Ensure a consistent CLI for all entrypoints that always has `output-path` as an input directory. This might create confusion, since CLI apps like `generate_metrics`, `construct_csv` and `recover_config` need to create a single file. For example, how do we handle the case where a user wants to pass different configurations to generate different metrics? 

From a software design perspective, I feel the "best practice" would be 2, but 1 works better from the perspective of usability. Thoughts, @VukW?

Reply:
I am not sure if it is a good idea. I am thinking about creating a new CLI command for logging. This way, if the user wants to store the logs, they can define the location. By default, we can flush the logs to the console.

Reply:
@benmalef The issue is we want to have logs by default, not only when user explicitly turns it on. The reason is user might need to see a logs of a long three-days training job on the cluster that failed - so he (1) don't want to rerun the whole job to get the error, (2) might not be able to see the console output after job was killed.

So the whole goal of this PR is actually to add a persistent logging to the file so user can always reach to. `debug-info`, `generate-metrics`, `construct-csv` are actually exclusions from the general behavior: it's ok if we would not have log files for these commands as the commands are quick enough and usually user sees its console output anyway.

Reply:
> How about this: let's default the location of the logs to None, in which case it flushes the logs to the console. In case the location is a directory, it writes the log there. This way, we don't have to change the invocation of generate_metrics, construct_csv and recover_config (which are relatively quick-running processes anyway).

🥇After thoughts, I believe, this your @sarthakpati  solution looks best to me. Maybe it's not so elegant from the point of code design, but it should be more comfortable and intuitive for users. For us it means we revert my previous proposal (where `gandlf_logger_setup` is called just once in `gandlf` cli command initialization) and instead run the configuration in the start of every "real" command code (right after cli is already called and cli params are already parsed)

🥈At the second option - same, but unify `--output-dir` param name, so we can extract it from the cli context & set up logging just once for all `gandlf` commands

Reply:
Agreed. We can have a follow-up PR where we specify `--log-file` and include this into the CLI apps. Sounds good, @VukW @benmalef?

Reply:
@sarthakpati Sounds good to me.

Reply:
> > How about this: let's default the location of the logs to None, in which case it flushes the logs to the console. In case the location is a directory, it writes the log there. This way, we don't have to change the invocation of generate_metrics, construct_csv and recover_config (which are relatively quick-running processes anyway).
> 
> 🥇After thoughts, I believe, this your @sarthakpati solution looks best to me. Maybe it's not so elegant from the point of code design, but it should be more comfortable and intuitive for users. For us it means we revert my previous proposal (where `gandlf_logger_setup` is called just once in `gandlf` cli command initialization) and instead run the configuration in the start of every "real" command code (right after cli is already called and cli params are already parsed)
> 
> 🥈At the second option - same, but unify `--output-dir` param name, so we can extract it from the cli context & set up logging just once for all `gandlf` commands


@VukW  If I get this right, these steps should be done? 
1. Change the ganldf_logger_setup()
```
def gandlf_logger_setup(log_dir=None)
   If (log_dir == None)
   # flush  the logs to the console 
   else:
  # store the logs in the log_dir
```
2. Call the `gandlf_logger_setup() ` in CLI "real" commands after parameters are parsed. 
For example, for the `collect_stats` command [ref](GANDLF/entrypoints/collect_stats.py): 

``` 
 args = parser.parse_args()
gandlf_logger_setup(args.outputdir)
 _collect_stats(args.modeldir, args.outputdir)
```

Reply:
I agree. However, let's do 2 in a separate PR so that it is cleaner.

Reply:
Hey folks!

> Agreed. We can have a follow-up PR where we specify --log-file and include this into the CLI apps. 

@sarthakpati , sorry, didn't get. Do you mean we add a new explicit param to specify log file path? It would be optional, right, and without default value? But in this case user who faces failure of his looongish training run won't have the log. I thought one of the main reasons for us to implement a better logging is to log debug details always (or, at least, by default) so user can debug his run in case of any issues?

Reply:
> It would be optional, right

Yup!

> But in this case user who faces failure of his looongish training run won't have the log.

For now, until @benmalef puts the new CLI param in place. I think this is unfortunately going to be a case where a user needs to be cognizant of the fact that the logs are going to be long. Considering other factors (write access, default write locations, so on), I think we only have 2 options for the default:

1. Flush to console.
2. Write to a temporary file using the [tempfile module](https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile).

I believe you would prefer 2, right, @VukW?

Reply:
I am fine with flushing to console, but I understand what @VukW is saying about it being a pain. Do you think changing the default behavior to use a [temporary file](https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile) would be quick, @benmalef? 

I am fine with 1 & 2 of your proposals, but 3 & 4 represent a significant departure from the current workflow of scientific computing software packages, and I feel needs discussion amongst the broader audience/developer group.

Reply:
@sarthakpati @VukW 
I tried to implement [here](https://github.com/benmalef/GaNDLF/pull/15)
Also, I tried to use the tempfile module, but I get some errors.

Reply:
@benmalef @sarthakpati 
concerning (3) I believe it's both ok to make init once in `cli_tool` if possible (it was actually created to unite some general stuff like logging), as well as init logs in a separate commands if it's necessary.

concerning (4) and `gandlf init` - I agree with Sarthak that this would require a loot of changes (like, config management  / versioning / init), so I'd prefer not to implement this right now.

Reply:
Thanks @VukW!

> concerning (3) I believe it's both ok to make init once in `cli_tool` if possible 

@benmalef I guess we can merge once this is done. Thanks!

Reply:
@sarthakpati  Perfect..!! Tomorrow I will start to implement the next PR (CLI commands)